csl/core/engine.lua

```diff
@@ -0,0 +1,967 @@
+      end
+   end
+   if matching then
+      return self:_render_children(content, entry), true
+      -- FIXME:
+      -- The CSL specification says: "Delimiters from the nearest delimiters
+      -- from the nearest ancestor delimiting element are applied within the
+      -- output of cs:choose (i.e., the output of the matching cs:if,
+      -- cs:else-if, or cs:else; see delimiter).""
+      -- Ugh. This is rather obscure and not implemented yet (?)
+   end
+   return nil, false
+end
+
+function CslEngine:_choose (options, content, entry)
+   for _, c in ipairs(content) do
+      if c.command == "cs:if" or c.command == "cs:else-if" then
+         local t, match = self:_if(c.options, c, entry)
+         if match then
+            return t
+         end
+      elseif c.command == "cs:else" then
+         return self:_render_children(c, entry)
+      end
+   end
+end
+
+function CslEngine:_sort (options, content, entry)
+   -- FIXME TODO
+   -- Silent for now.
+   -- SU.warn("CSL sort not implemented yet")
+end
+
+-- PROCESSING
+
+function CslEngine:_render_node (node, entry)
+   local callback = node.command:gsub("cs:", "_")
+   if self[callback] then
+      return self[callback](self, node.options, node, entry)
+   else
+      SU.warn("Unknown CSL element " .. node.command .. " (" .. callback .. ")")
+   end
+end
+
+function CslEngine:_render_children (ast, entry, context)
+   if not ast then
+      return
+   end
+   local ret = {}
+   context = context or {}
+   for _, content in ipairs(ast) do
+      if type(content) == "table" and content.command then
+         local r = self:_render_node(content, entry)
+         if r then
+            table.insert(ret, r) --.. callback)
+         end
+      else
+         SU.error("CSL unexpected content") -- Should not happen
+      end
+   end
+
+   return #ret > 0 and self:_render_delimiter(ret, context.delimiter)
+end
+
+function CslEngine:_postrender (text)
+   local rdquote = self.punctuation.close_quote
+   local ldquote = self.punctuation.open_quote
+   local rsquote = self.punctuation.close_inner_quote
+
+   -- Typography: Ensure there are no double straight quotes left from the input.
+   text = luautf8.gsub(text, "([%s%p])\"", "%1" .. ldquote)
+   text = luautf8.gsub(text, "\"([%s%p])", rdquote .. "%1")
+   -- HACK: punctuation-in-quote is applied globally, not just to generated quotes.
+   -- Not so sure it's the intended behavior from the specification?
+   if SU.boolean(self.locale.styleOptions["punctuation-in-quote"], false) then
+      text = luautf8.gsub(text, "([" .. rdquote .. rsquote .. "]+)%s*([.,])", "%2%1")
+   end
+   -- HACK: fix some double punctuation issues.
+   -- Maybe some more robust way to handle affixes and delimiters would be better?
+   text = luautf8.gsub(text, "%.%.", ".")
+   -- Typography: Prefer to have commas and periods inside italics.
+   -- (Better looking if italic automated corrections are applied.)
+   text = luautf8.gsub(text, "(</em>)([.,])", "%2%1")
```

Comment:
Suggested:

```
   text = luautf8.gsub(text, "(</em>)([%.,])", "%2%1")
   -- HACK: remove extraneous periods after exclamation and question marks.
   -- (Follows the preceding rule to also account for moved periods) 
   text = luautf8.gsub(text, "([!?])%.", "%1") -- Remove trailing punctuation
```

See https://github.com/sile-typesetter/sile/issues/2024#issuecomment-2198081835

Reply:
Not sufficient, we might also have to handled the quoted stuff when punctuation-in-quotes is not active: `« My Stuff! ».`

Real weird that CSL does not say a word about this...

Reply:
Being familiar with the state of localization tech when CSL was developed, I don't find that oversight surprising at all. I'm surprised it covers as many edge cases as it does (even if my eyes want to bleed looking at how it was done).

Reply:
Checking some (non-bibliography related) French typography resources, `« My Stuff ! ». Blah...` is _sometimes_ accepted...
So I'll be opinionated: It's ugly and hurts my feelings.[^1] 

I certainly prefer `« My Stuff ! » Blah...` (but a comma _would_ look ok to me, my only concern is the period... Interesting how tastes and usage do affect things that do not have a "definitive" rule or convention...). 

So I'll also fix it.

(And my eyes want to bleed too, lol)

[^1]: I am a big advocate of not using "double marking" when not needed. It goes beyond the topic here, but for instance, some people write: 
     > This French guy said "_bonjour_" ("hello")...

    That's double marking (italic for foreign words, quotes for quotations). In _most_ situations, I tend to prefer:
    > This French guy said _bonjour_ ("hello")...

    There's no ambiguity, so a "single marking" is sufficient and remains typographically sound... and IMHO, it's more elegant (in addition to saving a few characters). Well... Actually, when I am the author, there are some contexts (e.g. in footnotes) where I may even prefer:
    > He actually said: "_bonjour tout le monde,_ hello everybody"...

    (It also boils down to removing some double marking: a parenthesized translation)

Reply:
Besides ? and !, also handle suppression after an ellipsis...

Reply:
Fixed